README.md
ggfan
================
Jason Hilton
05 March, 2019

[![Build
Status](https://travis-ci.org/jasonhilton/ggfan.svg?branch=master)](https://travis-ci.org/jasonhilton/ggfan)

## Summarise a distribution through coloured intervals

This package is a ggplot-based implementation of some of the
functionality of the [`fanplot`
package](https://gjabel.wordpress.com/category/r/fanplot/) by Guy Abel.
Fanplot provides methods to visualise probability distributions by
representing intervals of the distribution function with colours.
Plotting samples from posterior distributions obtained through MCMC is a
particular aim. A ggplot-based implementation is useful as it allows
leverage of the power of [ggplot2](http://ggplot2.tidyverse.org/)
features allowing, for example, facetting and easy theming.

## Installation

Installation directly from CRAN should be possible shortly, once the
package has been resubmitted and checked: `install.packages("ggfan")`
Otherwise, the package can be installed directly from github using the
devtools package: `devtools::install_github("jasonhilton/ggfan")`.

## Quick Start

The provided `fake_df` data gives and example of the type of data you
might want to plot with ggfan: it consists of 1000 samples of an outcome
variable of interest at each value of a covariate `x`. We can plot this
simply using standard `ggplot2` syntax and `geom_fan`. Convenient
`ggplot` features such as themes, colour scales and facetting can also
be
used.

``` r
ggplot(fake_df, aes(x=x, y=y)) + geom_fan() + theme_minimal() + scale_fill_distiller(palette="Spectral")
```

![](tools/README-example-1.png)<!-- -->

## Further Information

A full description of the functionality of `ggfan` is provided in the
vignette, provided
[here](https://github.com/jasonhilton/ggfan/blob/master/vignettes/geom_fan.md).

A brief example of using `ggfan` to plot `stan` output is given in a
second vignette
[here](https://github.com/jasonhilton/ggfan/blob/master/vignettes/ggfan_stan.md)


NEWS.md
# ggfan 0.1.4
* Ensures vignette compatibility with new version of `rstan` (which now no longer imports ggplot2 on which I was stupidly relying). Thanks to Ben Goodrich for highlighting this (and for all the cool work done by the stan devs in general). 
* Fixes notes regarding unneeded imports.

# ggfan 0.1.3
* Ensures compatibility with updates of dplyr and tibble (thanks to Francois Romain for adding the relevant import for dplyr.)

# ggfan 0.1.2
* Ensures compatibility with ggplot 2.2.1.9000 and hopefully therefore 2.3.0.
* Adding a default linetype to GeomIntervalPath prevented an error in the call to draw_key_path in the new version of ggplot2.
* Fix documentation links to external packages

# ggfan 0.1.1

* Make `rstan` a strong dependency to avoid warnings when running vignette or on load of dataset if this package is not installed on test platform.

# ggfan 0.1.0

Initial submission to CRAN


tests/testthat.R
library(testthat)
library(ggfan)

test_check("ggfan")


tests/testthat/test_calc_quantile.R
context("calc_quantile")


test_that("calc_quantile returns df with x and y",
  {
  intervals <- c(0.5,0.8)
  df <- data.frame(x=rep(1:10,5), y=rnorm(50))
  df_q <- calc_quantiles(df, intervals)
  expect_true(all(c("x","y") %in% names(df_q)))
  }
)

test_that("calc_quantile allows retention of original df names",
  {
  intervals <- c(0.5,0.8)
  df <- data.frame(a=rep(1:10,5), b=rnorm(50))
  df_q <- calc_quantiles(df, intervals, x_var="a", y_var="b", rename=F)
  expect_true(all(c("a","b") %in% names(df_q)))
  }
)

test_that("calc_quantile returns error when too few observations are present",
  {
  intervals <- c(0.5,0.8)
  df <- data.frame(x=rep(1:10,2), y=rnorm(20))
  expect_error(df_q <- calc_quantiles(df, intervals))
  }
)



tests/testthat/test_stat_interval.R
context("stat_interval")

test_that("stat_interval warns when there are fewer observations than
           the quantiles to be plotted",{
             df <- data.frame(x=rep(1:10,2), y=c(rnorm(20)))
             expect_warning(ggplot2::ggplot_build(
                              ggplot2::ggplot(df, ggplot2::aes(x=x, y=y)) +
                            stat_interval(intervals = c(0.5,0.7,0.9))))

           })

test_that("mapping quantile aesthetic which are assymetric about median gives warning", {
          df <- data.frame(x=rep(1:10,2), quantile=c(rep(0.2,10), rep(0.6,10)),
                           y=c(rnorm(20)))
          expect_warning(ggplot2::ggplot_build(
            ggplot2::ggplot(df, ggplot2::aes(x=x,y=y,quantile=quantile)) +
                      stat_interval(intervals=c(0.2,0.6))))

})

test_that("stat_interval filters precomputed quantiles to match interval args",{ 
          df <-data.frame(x=rep(1:3,rep(4,3)), 
                          quantile=rep(seq(2,8,2)/10, 3),
                          y= rep(1:4,3))
          p_b <- ggplot2::ggplot_build(
            ggplot2::ggplot(df, ggplot2::aes(x=x,
                                             y=y,
                                             quantile=quantile)) +
                                       stat_interval(intervals=0.2))
          expect_equal(length(unique(p_b$data[[1]]$interval)), 1)
          expect_equal(unique(p_b$data[[1]]$interval), as.factor(0.2))
})

test_that("stat_interval give error with strings of quantiles",{ 
  df <-data.frame(x=rep(1:3,rep(4,3)), 
                  quantile=vapply(rep(seq(20,80,20), 3),
                                  function(x) paste0(x,"%"),
                                  FUN.VALUE ="4%"),
                  y= rep(1:4,3))
  expect_warning(ggplot2::ggplot_build(
    ggplot2::ggplot(df, ggplot2::aes(x=x, y=y, quantile=quantile)) +
                                 stat_interval(intervals=0.2)))
  
})

test_that("stat_interval throws error when given nonsensical quantiles",{ 
  df <-data.frame(x=rep(1:3,rep(4,3)), 
                  quantile=rep(seq(2,8,2)*100, 3),
                  y= rep(1:4,3))
  
  expect_warning(ggplot2::ggplot_build(
                   ggplot2::ggplot(df, 
                                   ggplot2::aes(x=x, y=y, quantile=quantile)) +
                 stat_interval(intervals=0.2)))
})

test_that("stat_interval gives error when precomputed quantiles and requested
           intervals don't match", {
             df <- data.frame(x=rep(1:10,2), quantile=c(rep(0.1,10), rep(0.9,10)),
                              y=c(rnorm(10,-3) ,rnorm(10,3)))
             expect_warning(ggplot2::ggplot_build(
               ggplot2::ggplot(df, ggplot2::aes(x=x,y=y,quantile=quantile)) +
                                    stat_interval(intervals=c(0.6))))
           })


R/ggfan.R
#' Fanplots for ggplot2
#'
#' Implements the functionality of the \code{fanplot} package as \code{ggplot}
#' geoms. Designed for summarising MCMC samples from a posterior distribution,
#' where a visualisation is desired for several values of a continuous
#' covariate.
#' Increasing posterior intervals derived from the quantiles of the sampled
#' quantity are mapped to a continuous colour scale.
#'
#'
#' @section Functions:
#' The package contains three plotting functions.
#'
#'
#' \code{\link{geom_fan}} produces fan plots
#'
#' \code{\link{geom_interval}} produces line plots, where pairs of lines represent
#' intervals
#'
#' \code{\link{stat_sample}} randomly plots a specified number of samples from the
#' data
#'
#' An additional function \code{\link{calc_quantiles}} computes relevant
#' quantiles for a specified set of intervals.
#'
#'
#' @docType package
#'
#' @name ggfan
#'
#'
NULL

## quiets concerns of R CMD check re: the .'s that appear in pipelines
if(getRversion() >= "2.15.1")  utils::globalVariables(c(".", "n" ,"x"))
# see https://github.com/tidyverse/magrittr/issues/29 and
# https://github.com/jennybc/googlesheets/commit/cf829306284601888035b2bad5988d6387187d6e


R/fake_df.R
#' Fake dataset intended to resemble a set of MCMC samples of a variable over
#' one covariate (perhaps time),
#'
#' The code needed to recreate the dataset is included in the examples
#'
#'
#' @format A data frame with 50000 rows and 3 columns
#' \describe{
#'  \item{x}{Values of the covariate}
#'  \item{y}{Values of the modelled quantity}
#'  \item{Sim}{Index referring to a MCMC posterior sample}
#' }
#'
#' @examples
#' # generate mean and variance for sequence of samples over time
#' library(magrittr)
#' library(tidyr)
#' set.seed(234)
#' N_time <- 50
#' N_sims <- 1000
#' time <- 1:N_time
#' mu <- time**2 * 0.03 + time * 0.3
#' sds <- exp(time**2 * -0.001 + time * 0.1)
#'
#' # simulate 1000 samples from each time point
#' fake_data <- sapply(time, function(i) rnorm(N_sims, mu[i], sds[i]))
#'
#' # gather into a long-form, tidy dataset
#' fake_df <- data.frame(x=time, t(fake_data)) %>%
#' tidyr::gather(key=Sim, value=y, -x)
#' # devtools::use_data(fake_df)
#'
"fake_df"



R/geom_fan.R
#' Fan plot visualising intervals of a distribution
#'
#' Fan Plots allow the distribution of a variable to be visualised by
#' representing sets of central probability intervals through colour.
#' For every value of \code{x}, \code{geom_fan} computes quantiles of \code{y} and uses these to plot
#' intervals containing increasing proportions of the total density of \code{y}.
#' Intervals are mapped to a continuous colour scale, so that changes in colour
#' represent intervals covering an increasing proportion of total density.
#' Quantiles can also be precomputed and mapped to the aesthetic \code{quantile}.
#' This function is designed with the need to summarise MCMC posterior
#' distributions in mind, and implements the functionality of the \code{fanplot}
#' package in \code{ggplot2}. Note that there should be enough observations of 
#' \code{y} at each \code{x} to allow estimation of the specified quantiles.
#'
#' @inheritParams ggplot2::layer
#' @param stat Use to override the default use of \code{\link{stat_interval}}
#' @param intervals specify the collection of intervals to be represented in the
#' fan.
#' @param ... other arguments passed on to \code{layer}. These are
#'   often aesthetics, used to set an aesthetic to a fixed value, like
#'   \code{color = "red"} or \code{size = 3}. They may also be parameters
#'   to the paired geom/stat.
#'
#'
#' @section Aesthetics:
#'
#' \code{geom_fan} understands the following aesthetics (required aesthetics are in bold):
#'
#' \itemize{
#' \item \strong{x}
#' \item \strong{y}
#' \item \code{alpha}
#' \item \code{group}
#' \item \code{quantile}
#' }
#'
#' @export
#'
#'
#' @seealso
#' \code{stat_summary} Summarises y at each value of x
#'
#' \code{stat_quantile} Uses quantile regression to predict quantiles
#'
#' \code{geom_interval} Plot intervals boundaries as lines
#'
#'
#' @examples
#'
#' # Basic use. The data frame must have multiple y values for each
#' # x
#' library(ggplot2)
#'
#' ggplot(fake_df, aes(x=x,y=y)) +geom_fan()
#'
#'
#' # use precomputed quantiles - reducing storage requirements.
#' intervals = 1:19/20
#' fake_q <- calc_quantiles(fake_df, intervals=intervals)
#' # intervals in geom_fan must be the same as used to compute quantiles.
#' ggplot(fake_q, aes(x=x,y=y, quantile=quantile)) +
#'  geom_fan(intervals=intervals)
#'
#'
#' # change the colour scale
#' ggplot(fake_df, aes(x=x,y=y)) + geom_fan() + scale_fill_gradient(low="red", high="pink")
#'
#'
geom_fan <- function (mapping = NULL, data = NULL, stat = "interval", position = "identity",
         show.legend = NA, inherit.aes = TRUE, intervals=(2:98)/100, ...)
{

  ggplot2::layer(data = data, mapping = mapping, stat = stat,
                 geom = GeomIntervalPoly, position = position,
                 show.legend = show.legend, inherit.aes = inherit.aes,
                 params = list(intervals=intervals, ...))
}


#' See \code{\link[ggplot2]{ggplot2-ggproto}}
#' @format NULL
#' @usage NULL
#' @export
GeomIntervalPoly <- ggplot2::ggproto("GeomIntervalPoly", ggplot2::Geom,
    required_aes = c("x", "y", "fill", "interval","hilo"),
    default_aes=ggplot2::aes(alpha=1),
    draw_group = function(data, panel_scales, coord) {
      n <- nrow(data)
      if (n <= 2) return(grid::nullGrob())
      # exclude median, avoiding floating point comparison problems.
      tol <- 1e-6
      data <- dplyr::filter(data, abs(interval - 0.0) > tol)
      data <- dplyr::group_by(data, x)

      data <- dplyr::mutate(data, ordering = hilo * x)

      data <- dplyr::arrange(data, hilo, ordering, -interval)

      data <- dplyr::ungroup(data)
      coords <- coord$transform(data, panel_scales)
      grid::polygonGrob(
        coords$x, coords$y,
        id= - data$interval * 100,
        #id=-1,
        default.units = "native",
        gp = grid::gpar(col = 0,
                        fill = scales::alpha(data$fill, data$alpha),
                        lwd = 0,
                        lty = 0
                )
       )
    },
    draw_key=ggplot2::draw_key_rect
)

#' Line plot visualising intervals of a distribution
#'
#' For every value of \code{x}, computes quantiles of \code{y} and uses these to plot
#' intervals containing increasing proportions of the total density of \code{y}.
#' Boundaries of intervals are mapped to \code{linetype}. Quantiles can also be
#' precomputed and mapped to the aesthetic \code{quantile}.
#' This function is designed with the need to summarise MCMC posterior
#' distributions in mind.
#'
#' @section Aesthetics:
#'
#' \code{geom_interval} understands the following aesthetics (required aesthetics are in bold):
#'
#' \itemize{
#' \item \strong{x}
#' \item \strong{y}
#' \item \code{quantile}
#' \item \code{group}
#' \item \code{colour}
#' \item \code{size}
#' }
#'
#' @export
#' @inheritParams ggplot2::layer
#' @inheritParams ggplot2::geom_line
#' @param stat Use to override the default use of \code{\link{stat_interval}}
#' @param intervals specify the collection of intervals to be represented in the
#' fan.
#'
#' @seealso
#' \code{stat_summary} Summarises y at each value of x
#' \code{stat_quantile} Uses quantile regression to predict quantiles
#' \code{geom_fan} Plot intervals on a continuous colour scale
#'
#'
#' @examples
#'
#' library(ggplot2)
#' # Basic use. The data frame must have multiple y values for each
#' # x
#' ggplot(fake_df, aes(x=x,y=y)) +geom_interval()
#'
#'
#' # use precomputed quantiles - reducing storage requirements.
#' intervals = c(0,50,90)/100
#' fake_q <- calc_quantiles(fake_df, intervals=intervals)
#' # intervals in geom_fan must be the same as used to compute quantiles.
#' ggplot(fake_q, aes(x=x,y=y, quantile=quantile)) +
#'  geom_interval(intervals=intervals)
#'
#'
geom_interval <- function(mapping = NULL, data = NULL,
                      stat = "interval_fctr", position = "identity",
                      intervals=c(0,50,90)/100,
                      lineend = "butt",
                      linejoin = "round",
                      linemitre = 1,
                      arrow = NULL,
                      na.rm = FALSE,
                      show.legend = NA,
                      inherit.aes = TRUE,
                      ...) {
  ggplot2::layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomIntervalPath,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      intervals=intervals,
      lineend = lineend,
      linejoin = linejoin,
      linemitre = linemitre,
      arrow = arrow,
      na.rm = na.rm,
      ...
    )
  )
}

#' See \code{\link[ggplot2]{ggplot2-ggproto}}
#' @format NULL
#' @usage NULL
#' @export
GeomIntervalPath <- ggplot2::ggproto("GeomIntervalPath", ggplot2::Geom,
  required_aes = c("x", "y", "interval","hilo","linetype"),

  default_aes = ggplot2::aes(colour = "black", size = 0.5, 
                             alpha = NA,
                             linetype="solid"),
  draw_group = function(data, panel_scales, coord, arrow = NULL,
                        lineend = "butt", linejoin = "round", linemitre = 1,
                        na.rm = FALSE) {

    data$line_id <- interaction(data$interval,data$hilo)

    data <- data[order(data$line_id), , drop = FALSE]
    coords <- coord$transform(data, panel_scales)

    grid::polylineGrob(
      coords$x, coords$y, id = data$line_id,
      default.units = "native", arrow = arrow,
      gp = grid::gpar(col = alpha(data$colour, data$alpha),
                lwd = data$size * .pt,
                lty = unique(data$linetype),
                lineend = lineend,
                linejoin = linejoin,
                linemitre = linemitre
        )
     )
  },
  draw_key = ggplot2::draw_key_path
)


R/gp_model_fit.R
#' A \code{stan_fit} object used in the \code{ggfan_stan} vignette, containing posterior samples
#' from a latent gaussian process model. This is provided as data to avoid 
#' having to conduct computationally expensive sampling when producing the vignettes.
#' 
#'
#' The code needed to recreate the object is included in the examples, as well 
#' as in the vignette code chunks. 
#'
#'
#' @format A `stan_fit` object containing samples of the following parameters.
#' \describe{
#'   \item{eta_sq}{Gaussian process variance parameter}
#'   \item{rho_sq}{Gaussian process roughness parameter}
#'   \item{z}{Latent poisson rate}
#'   \item{y_gen}{Posterior predictive sample of counts `y`} 
#' }
#' See the help page for \code{\link[rstan]{stanfit-class}} for more details.
#' 
#' 
#'
#' @examples
#' \dontrun{
#' # generate mean and variance for sequence of samples over time
#' library(rstan)
#' library(dplyr)
#' library(magrittr)
#' library(tidyr)
#' library(tibble)
#' 
#' library(ggfan)
#' seed <- 34526
#' set.seed(seed)
#' 
#' # data 
#' x <- seq(-5,5,0.1)
#' N <- length(x)
#' y <- cbind(rpois(N, exp(sin(x)+2)),rpois(N, exp(sin(x)+2)))
#' 
#' stan_data <- list(N=N, x=x, y=y)
#' 
#' 
#' compiled_model <- stan_model(file=file.path(path.package("ggfan"), 
#'                                             "stan","latent_gp_pois.stan"))
#' gp_model_fit <- sampling(compiled_model, data=stan_data, iter=3000,thin=6)
#' #devtools::use_data(gp_model_fit, internal=FALSE)
#' }
"gp_model_fit"

R/stat_interval.R
#' See \code{\link[ggplot2]{ggplot2-ggproto}}
#' @format NULL
#' @usage NULL
#' @importFrom dplyr n
#' @export
StatInterval <- ggplot2::ggproto("StatInterval", ggplot2::Stat,
  setup_params = function(data, params) {
    params
  },

  compute_group = function(data, scales, params, intervals) {
    precomputed_quantiles <- "quantile" %in% names(data)
    if(length(data$x)<2){
      stop("Too few rows in plotting data. This may be because invalid data has
           been provided")
    }

    # calculate quantiles if not already present.
    if (!precomputed_quantiles){
      data <- calc_quantiles(data, intervals)
    } else {
      if (!inherits(data$quantile, "numeric")){
        stop("Please ensure quantiles are inputed as a numeric vector with
             values bounded by 0 and 1")
      }
      if (max(data$quantile) > 1 | min(data$quantile) < 0){
        stop("Please ensure quantiles are inputed as a numeric vector with
             values bounded by 0 and 1")
      }
    }

    tol <- 1e-6
    data_interval <- dplyr::mutate(data,
                                   Interval = abs(quantile - 0.5) * 2)
    # round to avoid problems grouping by floating point.
    data_interval <- dplyr::mutate(data_interval, Interval=round(Interval, 3))

    if(precomputed_quantiles){
      # if the quantiles were precomputed they do not necessarily correspond
      # to intervals in the argument.
      # Filter to only contain those intervals requested
      data_interval <- dplyr::filter(data_interval,
                              .in_numeric(Interval, intervals, tol))
      if (dim(data_interval)[1] ==0){
        stop("No rows in plotting data frame. This may be because the intervals
              requested do not correspond to the precomputed quantiles you
              provided in the data. \n
              e.g. To plot interval 0.5, you need quantiles 0.25 and 0.75 \n
              Default intervals are 0, 0.5, and 0.9, so quantiles needed are
              c(0.05,0.25,0.5,0.75,0.95)")
      }
    }

    data_interval <- dplyr::group_by(data_interval, x, Interval)
    # exclude median, as is the empty interval (0.5,0.5).
    # Use tol. to avoid floating point wierdness.
    df_n <- dplyr::filter(data_interval, abs(quantile - 0.5) > tol)
    df_n <- dplyr::summarise(dplyr::group_by(df_n,Interval, x), n=n())
    if(!all(df_n$n==2)){
      stop(paste("Not all intervals have exactly two values at each x (high and low boundaries)",
                 "Note quantiles supplied must be in symmetric pairs about the median.",
                 "For example c(0.2,0.5,0.8) is admissable, c(0.1,0.5,0.51) is not"
      ))
    }
    data_interval <- dplyr::mutate(data_interval,
                                   hilo=ifelse(rank(y,ties.method="first")==n(),
                                               1, -1))
    data_interval <- dplyr::ungroup(data_interval)

    return(data_interval)
  },
  required_aes = c("x", "y"),

  default_aes = ggplot2::aes(fill=..Interval..,
                    interval=..Interval..,

                    hilo=..hilo..,
                    alpha=1)
)

#' Calculate quantiles of a tidy dataframe
#'
#' @param data  A data frame with containing x and y columns, with several y
#' values for every x
#' @param intervals A list of intervals for which corresponding quantiles are
#' desired.
#' @param x_var A character string giving the name of the x variable
#' @param y_var A character string giving the name of the y variable
#' @param rename Logical. Indicates whether to retain existing variable name
#' or use \code{x} and \code{y}.
#' @return A data frame containing x, y, and quantile columns (possibly renamed)
#'
#'
#' @examples
#'
#' head(fake_df)
#'
#' fake_q <- calc_quantiles(fake_df, intervals=c(0,0.5,0.8))
#' head(fake_q)
#'
#' @export
calc_quantiles <- function(data, intervals, x_var="x",y_var="y",rename=T){
  if (!(is.data.frame(data))){
    stop("data must be a data frame")
  }
  if (!(x_var %in% names(data) && y_var %in% names(data))){
     stop("specified x and y variables are not in dataframe")
  }

  probs <- c(rev(0.5 - intervals / 2), 0.5 + intervals / 2)
  # TODO recode with simpler way of doing this using rename_ or similar do_
  data$x <- data[[x_var]]
  data$y <- data[[y_var]]
  grouped_df <- dplyr::group_by(data, x)
  n_df <- dplyr::summarise(grouped_df, n=n())
  if (min(n_df$n) < length(probs)){
    stop("too few observations per x value to compute specifed intervals")
  }
  data_q <- dplyr::do(grouped_df,
                      data.frame(quantile=probs,
                                 y=stats::quantile(.$y, probs=probs)))
  if(!rename){
    data_q[[x_var]] <- data_q$x
    data_q[[y_var]] <- data_q$y
    data_q$x <- NULL
    data_q$y <- NULL
  }
  return(data_q)
}


#' See \code{\link[ggplot2]{ggplot2-ggproto}}
#' @format NULL
#' @usage NULL
#' @export
StatIntervalFctr <- ggplot2::ggproto("StatIntervalFctr", StatInterval,

 compute_group= function(data, scales, params, intervals){

    data <- StatInterval$compute_group(data,scales,params,intervals)
    data <- dplyr::mutate(data,
                          Interval_cont=Interval,
                          Interval=as.factor(Interval),
                          Int_group=-Interval_cont * hilo)
    int_levs <-levels(data$Interval)
    int_levs[int_levs=="0"] <- "Median"
    levels(data$Interval) <- int_levs
    return(data)
  },
  default_aes = ggplot2::aes(interval=..Interval..,
                    group=..Int_group..,
                    linetype=..Interval..,
                    hilo=..hilo..)
)

#' Line plot visualising intervals of a distribution
#'
#'  Very similar to \code{\link{geom_interval}}, except uses
#' \code{\link[ggplot2:geom_path]{geom_line}} to handle the plotting. This makes handling
#' plotting of intervals for several groups difficult to achieve, so
#' \code{\link{geom_interval}} is preferred.
#'
#' @inheritParams geom_interval
#' @export
stat_interval <- function (mapping = NULL, data = NULL, stat = "interval_fctr",
                           position = "identity", na.rm = FALSE,
                           show.legend = NA, inherit.aes = TRUE,
                           intervals=c(0,50,90)/100, ...)
{
  ggplot2::layer(data = data, mapping = mapping, stat = stat, geom = ggplot2::GeomLine,
        position = position, show.legend = show.legend,
        inherit.aes = inherit.aes,
        params = list(na.rm = na.rm, intervals=intervals, ...)
        )
}

#' See \code{\link[ggplot2]{ggplot2-ggproto}}
#' @format NULL
#' @usage NULL
#' @export
StatSample <- ggplot2::ggproto("StatSample", ggplot2::Stat,
  required_aes = c("x", "y", "group"),

  compute_panel= function(data, scales, params, n_samples=5){
    chosen_groups <- sample(data$group,n_samples)
    data <- data[data$group %in% chosen_groups,]
    return(data)
  },
  default_aes=ggplot2::aes(size=0.2)
)


#' Plots a randomly chosen sample of the specified groups using
#' \code{\link[ggplot2:geom_path]{geom_line}}
#'
#'
#' @param n_samples number of samples to plot
#' @param size The width of the line in mm
#' @param alpha The transparency of lines to be drawn. Must lie between 0 and 1.
#' @inheritParams  ggplot2::geom_line
#'
#'
#' @export
stat_sample <- function (mapping = NULL, data = NULL, stat = "sample",
                         position = "identity", na.rm = FALSE, show.legend = F,
                         inherit.aes = TRUE, n_samples=5, size=0.2, alpha=1.0,
                         ...)
{
  ggplot2::layer(data = data, mapping = mapping, stat = stat, geom = ggplot2::GeomLine,
        position = position, show.legend = show.legend,
        inherit.aes = inherit.aes,
        params = list(na.rm = na.rm,
                      n_samples=n_samples,
                      alpha=alpha,
                      size=size, ...)
                )
}


#' Find elements of one numeric vector in another.
#'
#' Find which elements of `a` are equal to at least one element in `b`, to
#' within some tolerance tol.
#'
#' @param a The vector for which comparison are to be made
#' @param b The vector to match against.
#' @param tol The tolerance within which values are assumed to be the same.
#'
#' @return A vector of logical values the same length as `a` describing whether
#' this vector element is close to at least one element in `b`
#'
.in_numeric <- function(a,b, tol=1e-6){
  if (length(b)==1){
    return(abs(a-b) < tol)
  } else{
    logical_mat <- vapply(a, function(x,y) abs(x-y) < tol,
                    FUN.VALUE=logical(length = length(b)),b)
    logical_vec <- apply(logical_mat, 2, any)
  }
  return(logical_vec)
}
